### ISSUE TITLE
Support contiguous tensor views with shared storage

### ISSUE THREAD

**Issue by isaacperez** [isaacperez/tabgrad#86]
## Problem

At `0abaee34e8cba351d22d5d01518001dda49256c4`, Tabgrad supports scalar and multidimensional contiguous CPU float32 values, equal-shape addition and ordinary observation. It cannot expose a second tensor shape over the same logical storage. Runtime materialization and CPU retirement currently associate ownership with individual values/slots; adding another handle with different metadata alone would not establish safe storage sharing.

## Why it matters

Callers need to reinterpret tensor dimensions without copying data. A bounded real view exercises the distinction between tensor identity, logical value and storage before broader layouts, mutation or differentiation build on it. The result must preserve memory reuse and lazy execution rather than adding a shape-only frontend shortcut.

## Expected result

Provide shape-only `Tensor.view` for contiguous CPU float32 tensors through Python and JavaScript. A view has its own immutable shape and handle, shares its source's logical storage, preserves flat element order and can feed existing equal-shape addition and normal observation. Creating a view does not execute pending arithmetic, allocate numerical payload or copy it. Sharing remains correct when handles close and when execution completes or fails.

## Included and excluded work

Include Python `Tensor.view(*shape)` with positional built-in integer dimensions or one built-in tuple/list of dimensions, including an empty sequence for a scalar; JavaScript `Tensor.view(shape)` with a dimension array; one inferred `-1` dimension when unambiguous; zero-sized and singleton dimensions; and the existing runtime representability limits. Frontends normalize syntax, while shared runtime admission owns normalized shape validity and storage semantics. Invalid dimensions, incompatible element counts, multiple inferred dimensions and ambiguous empty inference fail at admission. The pinned oracle establishes exact covered Python error categories before implementation; unsupported overloads fail explicitly.

Include host-backed, pending-computed and resident sources, sibling and chained views, existing observation and addition, and the ownership changes necessary across semantic state, materialization, program formation, CPU invocation storage and the Python bridge. Start inspection from those owners and their tests, not a repository-wide redesign.

Exclude dtype-reinterpretation `view(dtype)`, `reshape` with copy fallback, `view_as`, transpose/permute, slicing/indexing, offsets and arbitrary strides, broadcasting, mutation, autograd, new numerical operations/types/devices, a public storage-pointer or diagnostic API, dependencies and package publication. These remain route capabilities rather than placeholder subissues. Any newly discovered material contract choice returns to the research/architecture process before dependent implementation.

## Completion conditions

- [ ] Supported Python and JavaScript calls preserve element order/count and independent shape identity for scalar, vector, higher-rank, singleton and empty cases. Invalid calls fail synchronously without creating numerical work or leaking owners. Unsupported overloads are not silently reinterpreted.
- [ ] Views of host-backed, pending and resident values produce correct results directly and through existing addition. Closing the base does not invalidate a live view; closing a view does not invalidate its base or sibling. Chained views and repeated operands remain correct.
- [ ] Source and view demands share materialization of the same storage: a metadata view does not add a numerical kernel, duplicate upload/allocation or force an intermediate readback. Ordinary observation can still copy results into caller-owned output containers. Establish sharing with internal ownership/allocation evidence, not numerical equality alone and not a new public pointer API.
- [ ] Logical retention and physical last use are accounted for at their proper owners. Retained aliases, pending consumers and accepted observations prevent premature release/reuse; closing the last owner releases storage exactly once. A view over a computed value remains usable after completed producer ancestry is detached. Failure rollback preserves borrowed/live inputs and does not double-release aliases. Session close drains/releases all owned resources.
- [ ] Existing scalar/SIMD arithmetic, iterative graph formation/release and ownership-safe scratch reuse remain valid. Mixed view/arithmetic chains do not permanently retain obsolete view or producer history. Deep metadata chains have stack-independent traversal/release, with costs proportional to live required structure rather than repeated scans of unrelated aliases or session history.
- [ ] Admission uses compact rank/argument metadata, not tensor-element walks. Bounded measurements distinguish view creation, first materialization and later observation; report payload allocations/copies, live/peak storage and repeated cleanup. Compare unchanged contiguous addition against the exact base, vary payload size at fixed rank, and vary a small number of views over one storage. Explain host bookkeeping and any material regression separately from numerical work.
- [ ] Python compatibility evidence comes from the pinned native oracle, while browser integration covers the managed Python path and direct JavaScript in the configured Chrome/Firefox environments. No `_async()` Python observation method, mandatory JSPI or numerical host fallback is introduced.
- [ ] The independent design review traces future non-contiguous access and saved-value/mutation ownership as established extension axes: identify additions, modifications and stable owners without implementing them. Any changed operand enumeration keeps lowering, use counts and physical retirement coherent; metadata views must not be faked as binary arithmetic. Document the actual contracts and support boundary in their primary homes.

Production behavior follows the existing TDD and independent verification/review rules. Definition of this issue is not executed compatibility or performance evidence.

Measurements are sequential, with one active workload/browser, no worker fan-out or all-core stress, at most 1,048,576 elements per tensor, 8 MiB of experiment-owned numerical payloads and 64 MiB of Tabgrad-owned CPU linear memory. Use bounded repetitions/watchdogs and stop on unexpected growth or resource pressure. Interpreter/browser memory is separate. Deep-chain correctness uses tiny payloads; no model benchmark or dependency installation is required.

## Dependencies and related work

- #79 is the current planning index, not an executable parent. This selects a bounded C04/C01 H1 result and exercises C03, C08, C10–C11 and C22–C23. It does not complete general views or change capability priorities. Delivery must report the demonstrated scope and planning impact.
- #82 / PR #83 supplied contiguous rank support. #22 established semantic view/version/lifetime constraints through historical research, not a production view implementation. Both are completed predecessors, not duplicate work or open blockers.
- #67 and #76 / PR #78 establish the logical-retention/physical-last-use boundary to preserve. #65 / PR #68, #66 / PR #72 and #71 / PR #75 supply reclamation and iterative traversal guarantees.
- #84 / PR #85 strengthen extension-cost review and record binary operand enumeration as a scoping concern, not a mandate for a separate general refactor.

Search enumerated all 85 open/closed issue and PR records on 2026-09-23, compared titles and body references to views, reshape, transpose, strides and aliasing, and inspected the relevant research and production exclusions. The only open record is #79; no open PR or dedicated production-view issue exists. Related research/implementation is completed. There is no executable parent, required subissue or unresolved blocking relationship to add.

An independent read-only issue-boundary check by `view_issue_scope` inspected the complete proposal, accepted semantic/admission/reuse contracts and affected runtime/CPU ownership at the named base. It found the result suitable for a bounded feature rather than renewed architectural research, conditional on the sharing/lifetime evidence above. It did not repeat the coordinator's history search or execute tests; this is issue-definition evidence, not implementation preflight or verification.

## Effects and risks

Governing sources are `docs/architecture/semantic-state.md`, `operation-admission.md` and `cpu-intermediate-reuse.md`: tensor identity is not storage identity; metadata views do not copy payload; semantic liveness and actual physical completion jointly permit reuse. Shared records own shared facts. Exact class layout and lowering representation remain implementation choices within these accepted boundaries, not a new engine or speculative framework.

The contiguous-only alternative isolates shared ownership while retaining existing flat kernels. Transpose would additionally require non-contiguous execution and observation; copy-based reshaping would not test the required sharing. Neither is included. The result is one integrated feature because frontend-only shape changes without shared execution/lifetime would not satisfy it.

Risks include double release, premature reuse when one alias dies, duplicated materialization, accidental producer retention, shape inference overflow and empty ambiguity, hidden per-view payload copies, and confusion between logical uses and physical storage uses. Preserve existing structured errors and distinguish oracle behavior from Tabgrad representation limits. New material limits or architecture changes require explicit decisions.

This extends the public APIs without changing existing calls, deployment, network data flow, dependencies or licenses. It does not establish mutation/gradient or non-contiguous compatibility. Explain shared-storage meaning for programmers new to tensor runtimes in concept documentation; concrete ownership and lifecycle go in component/flow documentation, and exact `view` syntax/errors/examples in its operation-specific reference. General documents contain reusable conclusions, not an implementation diary or operation-specific catalogue. Update compatibility evidence and relevant indexes alongside the production change.

## User or developer need

Reshape an existing contiguous value for another calculation without paying for duplicate tensor data, and preserve its usability independently of the original handle's lifetime.

## Example of the intended behavior

Inside the managed Python entry, after implementation:

```python
import torch

base = torch.tensor([1, 2, 3, 4, 5, 6], dtype=torch.float32)
matrix = base.view(2, 3)
assert base.shape == (6,)
assert matrix.shape == (2, 3)
assert (matrix + matrix).tolist() == [[2.0, 4.0, 6.0], [8.0, 10.0, 12.0]]
assert matrix.view(-1).shape == (6,)
```

The float literals describe this float32 example, not all future tensor types. Existing tensor/session cleanup remains required. JavaScript uses `base.view([2, 3])`, existing `.add()` and asynchronous flat `.toArray()`; Python keeps ordinary `.tolist()`.

## Affected public interface

New shape-only Python and JavaScript `Tensor.view` methods and their checked bridge contract. Existing `.shape`, addition, observation and handle/session cleanup must work with views. No raw storage, stride inspection or new observation endpoint is required by this increment.

## PyTorch compatibility reference

[PyTorch 2.14 Tensor.view](https://docs.pytorch.org/docs/2.14/generated/torch.Tensor.view.html) establishes shared data, shape compatibility, inferred dimensions and the separate dtype overload. This issue deliberately covers only shape changes of contiguous CPU float32 tensors. Use the repository-selected oracle to establish covered overload syntax, empty/scalar inference and error behavior; reference prose alone is not executable compatibility evidence. The JavaScript interface is Tabgrad's own contract.

## Deliberately deferred behavior

Non-contiguous views, index selection, mutation/version transitions, differentiation, broadcasting and further operation families remain in #79. They are relevant extension axes, not implied support or independently scheduled work from this issue.
